### ISSUE TITLE
Additional issues with path and routing

### ISSUE THREAD

**Issue by youming-lin** [IBM-Swift/Kitura#821]
I've tested a permutation of routes with subrouting, parameters, modifiers, and custom matches, and have found a few cases where Kitura's behavior deviates from that of Node Express (or when Express is also wrong, from the expected).

For all permutations tested, these are the common code:

Kitura:

``` swift
import Foundation
import SwiftyJSON
import Kitura
import HeliumLogger
import LoggerAPI

Log.logger = HeliumLogger()

let subrouter = Router()
let router = Router()

// ../helloworld
subrouter.all("/helloworld") { req, res, next in
    try res.send("Hello world").end()
}

Kitura.addHTTPServer(onPort: 3000, with: router)
Kitura.run()
```

Express:

``` js
'use strict';

var express = require('express');
var app = express();
var router = express.Router();

router.all('/helloworld', function(req, res, next) {
    res.send('Hello world');
});

app.listen(3000, function () {
  console.log('Example app listening on port 3000!');
});
```
#### Mounting subrouter on `""`

Kitura:

``` swift
// localhost:3000/helloworld -> 404
router.all("", middleware: subrouter)
```

Express:

``` js
// localhost:3000/helloworld -> Hello world
app.use('', router);
```

Kitura should return `Hello world` but returns `404` instead.
#### Mounting subrouter on `"/"`

Kitura:

``` swift
// localhost:3000/helloworld -> 404 -- ERROR
router.all("/", middleware: subrouter)
```

Express:

``` js
// localhost:3000/helloworld -> Hello world
app.use('/', router);
```

Kitura should return `Hello world` but returns `404` instead.
#### Mounting subrouter on custom match parameter with `?` modifier

Kitura:

``` swift
// localhost:3000/helloworld -> hangs, Router.swift:169 -- ERROR
// localhost:3000/123/helloworld -> Hello world
// localhost:3000/abc/helloworld -> hangs, Router.swift:169 -- ERROR
router.all("/:id(\\d+)?", middleware: subrouter)
```

Express:

``` js
// localhost:3000/helloworld -> Hello world
// localhost:3000/123/helloworld -> Hello world
// localhost:3000/abc/helloworld -> 404
app.use('/:id(\\d+)?', router);
```

Kitura should match Express’s responses in this case instead of throwing an error and then not responding to the request.
#### Mounting subrouter on custom match parameter with `*` modifier

Kitura:

``` swift
// localhost:3000/helloworld -> 404 -- ERROR
// localhost:3000/123/helloworld -> Hello world
// localhost:3000/123/456/helloworld -> 404 -- ERROR
// localhost:3000/abc/helloworld -> 404
router.all("/:id(\\d+)*", middleware: subrouter)
```

Express:

``` js
// localhost:3000/helloworld -> 404 -- ERROR
// localhost:3000/123/helloworld -> 404 -- ERROR
app.use('/:id(\\d+)*', router);
```

Express fails in this test as well. `*` should match 0 or many, not just 1.

**Comment by youming-lin** [comment]
Most of the issues in the OP has been fixed by #845. However, I've found an additional issue with `allowPartialMatch`:

Kitura:

```swift
// localhost:3000/helloworld -> 404 -- ERROR
// localhost:3000/123/helloworld -> 404
router.all("/:id", allowPartialMatch: false, middleware: subrouter)
```

Express:

```js
// localhost:3000/helloworld -> Hello world
// localhost:3000/123/helloword -> 404
app.all('/:id', router);
```

Kitura should match Express’s responses in this case.

Functionally speaking, Express's `.all(middleware)` should be similar to Kitura's `.all(... allowPartialMatch: false, middleware: middleware)`, whereas Express's `.use(middleware)` should be similar to Kitura's `.all(... allowPartialMatch: true, middleware: middleware)`.
